### PR TITLE
initalise repository with sphinx extensions

### PR DESCRIPTION
**Pull request by oliver-sanders**
Migrate Sphinx extensions here from the `cylc/cylc-doc` and `metomi/rose` projects.

* Theme extensions (css/js stored under `_static`) converted to Sphinx extensions.
  * Read The Docs theme modifications have been extracted into their own package for ease of portability.
  * Hieroglyph theme modifications have been extracted into their own package for ease of portability.
* Each extension is now a Python package.
  * The static resources (css/js) are now contained in this package.
  * These resources are registered with Sphinx via an absolute path.
* Extensions packaged under `cylc.sphinx_ext`.
* Auto-document extensions from their docstrings.
* Migrate tests.

The extension code has been reviewed in previous PRs, each extension is now documented with examples.

This repository is self documenting (run `make html`), the docs generated include explanations of each extension and example usage.

Where to put the built documentation i'm not so sure, gh-pages would be the obvious choice but I'm not sure how that would work with our Cylc organisation setup. Will push this to https://github.com/cylc/cylc-sphinx-extensions/issues/1

### CODE REVIEW

**Comment by oliver-sanders** [comment]
https://github.com/metomi/rose/pull/2374 is a reference for usage.

File: cylc/sphinx_ext/rtd_theme_addons/__init__.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+Modifications:
+
+1. Permit nested admonitions of differing class
+
+   .. note::
+
+      .. error::
+
+         The inner admonition should appear in a differet colour to the outer
```
Comment (kinow) [inline cylc/sphinx_ext/rtd_theme_addons/__init__.py]: s/differet/different/

File: cylc/sphinx_ext/rtd_theme_addons/__init__.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+   .. deprecated:: 1.2.3
+
+      Oh no!
+
+4. Make versionchanged items more prominent.
+
+   .. versionchanged:: 1.2.3
+
+      Someting changed!
```
Comment (kinow) [inline cylc/sphinx_ext/rtd_theme_addons/__init__.py]: s/Someting/Something

File: cylc/sphinx_ext/practical/admonitions.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+# -----------------------------------------------------------------------------
+"""An extension providing a directive for writing pratical sections."""
```
Comment (kinow) [inline cylc/sphinx_ext/practical/admonitions.py]: s/pratical/practical

File: cylc/sphinx_ext/practical/admonitions.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+from docutils import nodes
+from docutils.parsers.rst.directives.admonitions import BaseAdmonition
+
+
+class Practical(BaseAdmonition):
+    """Directive for practical sections in documentation.
+
+    This class serves as a standin for maintainability purposes. It is
```
Comment (kinow) [inline cylc/sphinx_ext/practical/admonitions.py]: s/standin/stand-in (I think)

File: cylc/sphinx_ext/practical/admonitions.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    equivalient to:
```
Comment (kinow) [inline cylc/sphinx_ext/practical/admonitions.py]: s/equivalient/equivalent

File: cylc/sphinx_ext/practical/admonitions.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+
+class PracticalExtension(Practical):
+    """Directive for practical extension exercises."""
+    NAME = 'Practical Extension'
+    CLASSES = ['note', 'spoiler']
+
+
+class Spoiler(BaseAdmonition):
+    """Directive for auto-hiden "spoiler" sections.
```
Comment (kinow) [inline cylc/sphinx_ext/practical/admonitions.py]: s/hiden/hidden

File: cylc/sphinx_ext/diff_selection/_static/js/diff_selector.js [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ * ------------------------------------------------------------------------- */
+
+/* Return a list of matches for the provided string, see also String.indexOf. */
+String.prototype.indiciesOf = function (sub) {
```
Comment (kinow) [inline cylc/sphinx_ext/diff_selection/_static/js/diff_selector.js]: nitpick.. I think it'd be indicesOf or indexesOf? (a few occurrences)

File: cylc/sphinx_ext/sub_lang/lexer.py [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+# -----------------------------------------------------------------------------
+"""An extension providing a lexer which highlights text <substitutions>."""
+
+from pygments.lexer import RegexLexer
+from pygments.token import Text, String, Comment
+
+
+class SubstitutionLexer(RegexLexer):
+    """Pygments lexer for highlighting <subsitutions> in code e.g. paths."""
```
Comment (kinow) [inline cylc/sphinx_ext/sub_lang/lexer.py]: s/subsitutions/substitutions

File: cylc/sphinx_ext/cylc_lang/lexers.py [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+        ],
+
+        # Parameterised syntax:  <foo=1>
+        'parameterisation': [
+            (r'(\<)'  # Opening greater-than bracket.
+             r'(\s?\w+\s?'  # Parameter name (permit whitespace).
+             r'(?:[+-=]\s?\w+)?'  # [+-=] for selecting parameters.
+             r'\s?'  # Permit whitespace.
+             r'(?:'  # BEGIN optional extra parameter groups...
+             r'(?:\s?,\s?\w+\s?'  # Comma seperated parameters.
```
Comment (kinow) [inline cylc/sphinx_ext/cylc_lang/lexers.py]: s/seperated/separated

File: cylc/sphinx_ext/minicylc/_static/js/minicylc.js [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+var minicylc_default_theme = {
+    'waiting_fill': 'none',
+    'waiting_stroke': '#88c6ff',
+    'running_fill': '#00c410',
+    'running_stroke': 'black',
+    'succeed_fill': '#ada5a5',
+    'succeed_stroke': 'black'
+}
+
+// Demo colour theme for demmonstrating workflow logic.
```
Comment (kinow) [inline cylc/sphinx_ext/minicylc/_static/js/minicylc.js]: s/demmonstrating/demonstrating

File: cylc/sphinx_ext/minicylc/_static/js/minicylc.js [hunk 11]
```diff
@@ -0,0 +1,310 @@
...
+                        '", defaulting to black.');
+            this.theme = {};
+        }
+    }
+
+    _find_svg_elements(svg) {
+        /**
+         * Associate task/dependency names with SVG nodes.
+         *
+         * Associations stored as dicionaries this.nodes and this.edges.
```
Comment (kinow) [inline cylc/sphinx_ext/minicylc/_static/js/minicylc.js]: s/dicionaries/dictionaries

File: cylc/sphinx_ext/minicylc/_static/js/minicylc.js [hunk 12]
```diff
@@ -0,0 +1,310 @@
...
+         * @param svg The <svg> element containg the workflow.
```
Comment (kinow) [inline cylc/sphinx_ext/minicylc/_static/js/minicylc.js]: s/containg/containing

File: cylc/sphinx_ext/minicylc/_static/js/minicylc.js [hunk 13]
```diff
@@ -0,0 +1,310 @@
...
+                changed = true;
+            }
+        }
+        return changed;
+    }
+
+    _main_loop(itt) {
+        /*
+         * The main loop - runs the simulation and handles restyling of nodes.
+         * Note function calls itself reccursively.
```
Comment (kinow) [inline cylc/sphinx_ext/minicylc/_static/js/minicylc.js]: s/reccursively/recursively

File: cylc/sphinx_ext/minicylc/minicylc.py [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+                if itt < len(deps) - 1:
+                    ors = cls.extract_or_deps(dep)
+                    # There is a RHS to this dep.
+                    for left in (i for i in dep if i not in
+                                 cls.CONDITIONAL_CHARS):
+                        for right in (i for i in deps[itt + 1] if i not in
+                                      cls.CONDITIONAL_CHARS):
+                            trigs.add((left, right, left in ors))
+                elif len(dep) == 1:
+                    # No depedent task.
```
Comment (kinow) [inline cylc/sphinx_ext/minicylc/minicylc.py]: s/depedent/dependent

File: cylc/sphinx_ext/practical/_static/js/spoiler.js [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+            $('<button />')
+                .addClass('btn')
+                .addClass('btn-' + button_class)
+                .addClass('spoiler-show')
+                .append('Show')
+        );
+
+    }
+
+    // On-click un-hide the contents of the adminition and hide the button.
```
Comment (kinow) [inline cylc/sphinx_ext/practical/_static/js/spoiler.js]: s/adminition/admonition (I think)

File: cylc/sphinx_ext/minicylc/minicylc.py [hunk 16]
```diff
@@ -0,0 +1,250 @@
...
+
+    """
+    option_spec = dict(GraphvizSimple.option_spec)
+    option_spec['snippet'] = directives.flag
+    option_spec['theme'] = directives.unchanged
+    option_spec['size'] = directives.unchanged
+    required_arguments = 0  # Arg will be provided in run().
+
+    CONDITIONAL_CHARS = ['&', '|', '(', ')']
+    CONDITIONAL_REGEX = re.compile(r'([\(\)\|\&])')
```
Comment (kinow) [inline cylc/sphinx_ext/minicylc/minicylc.py]: unnecessary escapes I think? `r'([\(\)\|\&])'` same as `re.compile(r'([()|&])')` ?

File: cylc/sphinx_ext/minicylc/__init__.py [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+.. minicylc::
+
+   foo => bar => baz
+
+"""
+
+
+from cylc.sphinx_ext.minicylc.minicylc import MiniCylc, MiniCylcDirective
+
+import sphinx
```
Comment (kinow) [inline cylc/sphinx_ext/minicylc/__init__.py]: Unused import

**Comment by oliver-sanders** [comment]
> Then tried running some of the commands listed in make help output (after installing via pip install -e .). But the commands I tried failed

I've packaged dependencies a little strangely so that you don't have to install dependencies for all extensions to install some of the extensions. `pip install -e .` just installs the base dependencies.

To install dependencies for a specific extension do:

```console
$ pip install -e [extension_name, ...]
```

To install everything:

```console
$ pip install -e .[all]
```

File: cylc/sphinx_ext/practical/admonitions.py (hunk 6 above)
Comment (oliver-sanders) [reply cylc/sphinx_ext/practical/admonitions.py]: Sorry for making you correct these twice!

**Comment by oliver-sanders** [comment]
@kinow Sorry for not copying the fixes from cylc-doc.

**Comment by kinow** [comment]
>I've packaged dependencies a little strangely so that you don't have to install dependencies for all extensions to install some of the extensions. pip install -e . just installs the base dependencies.

It was me being slow to read the README and check setup.py :+1: looking good, and with namespace and all.

**Comment by oliver-sanders** [comment]
@wxtim volunteering you for you packaging experience. The code itself is old (moved from rose) so doesn't require much attention, the packaging, docs, setup etc is new and needs review.

**Comment by wxtim** [comment]
> @wxtim volunteering you for you packaging experience. The code itself is old (moved from rose) so doesn't require much attention, the packaging, docs, setup etc is new and needs review.

Seen - Now looking

**Comment by wxtim** [comment]
Fancy adding a license?

**Comment by oliver-sanders** [comment]
> Fancy adding a license?

Licence now on `master`, referenced in `setup.py` in this branch.

**Pull request merged**